### ISSUE TITLE
I faced the biuld error of Libtorch's mnist.cpp in Docker container.

### ISSUE THREAD

**Issue by yoshihingis** [pytorch/examples#674]
(1) There is minor error in docker file
     I downloaded Dockerfile of  libtorch from https://github.com/freemanix/torchbuilder/blob/master/Dockerfile.
   MaybeI think that there is minor error as blow

 #_Prepare built package
RUN cd && mkdir -p libtorch/include && mkdir -p libtorch/share
(Right )RUN cp -r pytorch/build/build/lib libtorch
※(wrong) RUN cp -r pytorch/build//lib libtorch
RUN cp -r pytorch/torch/share/cmake libtorch/share/cmake
RUN for dir in ATen c10 caffe2 torch; do cp -r pytorch/torch/include/$dir libtorch/include; done
CMD [ "/bin/bash", "-" ]

(2)The Dockerfile was build successfully 
  I correct Dockerfile as above and the Dockerfile was build successfully .
 oshiki@kanna:~/Libtorch/test$ docker build -t libtorch:0 .
Sending build context to Docker daemon  6.144kB
Step 1/27 : FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu18.04
 ---> a387f44414ed
Step 2/27 : ENV http_proxy http://my.company.proxy
 ---> Using cache
 ---> bde22dd36e80
Step 3/27 : ENV https_proxy https://my.company.proxy
 ---> Using cache
 ---> c8b62fd013b2
Step 4/27 : USER root
 ---> Using cache
 ---> 8efd2ba5115b
Step 5/27 : WORKDIR /root
 ---> Using cache
 ---> 66aca9848ab3
Step 6/27 : RUN apt-get -y update && apt-get -y upgrade
 ---> Using cache
 ---> d9ca16c7fe54
Step 7/27 : RUN apt-get -y install g++ git libgflags-dev libgoogle-glog-dev    libiomp-dev libopenmpi-dev protobuf-compiler    python3 python3-pip python3-setuptools python3-yaml wget    ca-certificates
 ---> Using cache
 ---> 2e8b994a75a8
Step 8/27 : COPY /cert /usr/local/share/ca-certificates
 ---> Using cache
 ---> 2dc77ea21d14
Step 9/27 : RUN update-ca-certificates -f
 ---> Using cache
 ---> 278da53e2d3e
Step 10/27 : ARG CMAKE=cmake-3.14.1.tar.gz
 ---> Using cache
 ---> 7727f9114cb4
Step 11/27 : RUN wget https://github.com/Kitware/CMake/releases/download/v3.14.1/${CMAKE}
 ---> Using cache
 ---> a1271ac3d4be
Step 12/27 : RUN tar xvzf ${CMAKE} && cd cmake* && ./bootstrap --parallel=$(nproc)
 ---> Using cache
 ---> 7962a1af3f73
Step 13/27 : RUN cd cmake* && make -j$(nproc) && make install
 ---> Using cache
 ---> e0541a18be61
Step 14/27 : RUN wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB
 ---> Using cache
 ---> fb23dd6216ba
Step 15/27 : RUN apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB && rm GPG-PUB*
 ---> Using cache
 ---> 8b691ec1efcf
Step 16/27 : RUN sh -c 'echo deb https://apt.repos.intel.com/mkl all main > /etc/apt/sources.list.d/intel-mkl.list'
 ---> Using cache
 ---> b6f8e19f1362
Step 17/27 : RUN apt-get update && apt-get -y install intel-mkl-64bit-2019.1-053
 ---> Using cache
 ---> daf44f6a0f15
Step 18/27 : RUN rm /opt/intel/mkl/lib/intel64/*.so
 ---> Using cache
 ---> 8e9b99514120
Step 19/27 : ENV TORCH_CUDA_ARCH_LIST="5.2 6.0 6.1 7.0 7.5"
 ---> Using cache
 ---> 48cfc3317f91
Step 20/27 : ENV TORCH_NVCC_FLAGS="-Xfatbin -compress-all"
 ---> Using cache
 ---> 2a13a3ece008
Step 21/27 : RUN git clone --recurse-submodules -j8 https://github.com/pytorch/pytorch.git
 ---> Using cache
 ---> 0cb0a0e243fa
Step 22/27 : RUN cd pytorch && mkdir build && cd build && BUILD_TEST=OFF USE_NCCL=OFF python3 ../tools/build_libtorch.py
 ---> Using cache
 ---> 5a136c108634
Step 23/27 : RUN cd && mkdir -p libtorch/include && mkdir -p libtorch/share
 ---> Using cache
 ---> 7555af34ede3
Step 24/27 : RUN cp -r pytorch/build/build/lib libtorch
 ---> Using cache
 ---> 4d22b57db733
Step 25/27 : RUN cp -r pytorch/torch/share/cmake libtorch/share/cmake
 ---> Using cache
 ---> 96a44921852f
Step 26/27 : RUN for dir in ATen c10 caffe2 torch; do cp -r pytorch/torch/include/$dir libtorch/include; done
 ---> Using cache
 ---> 9bc74d571cf1
Step 27/27 : CMD [ "/bin/bash", "-" ]
 ---> Using cache
 ---> 54a57f97604f
Successfully built 54a57f97604f
Successfully tagged libtorch:0 

(3)$docker run -it --gpus all libtorch:0 /bin/bash to attach Docker container.

(4)there is the directory libtorch ,so $ cd libtorch
(5)$mkdir mnist and $mkdir build and $cd libtoch
(6) I copied CMaleList.txt and mnist.cpp from https://github.com/pytorch/examples/tree/master   /cpp/mnist

(7)$ cd build and $cmake -DCMAKE_PREFIX_PATH=/libtorch ..
 root@460b2c0da7ef:~/libtorch/mnist/build# cmake -DCMAKE_PREFIX_PATH=/libtorch ..
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:4 (find_package):
  By not providing "FindTorch.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Torch", but
  CMake did not find one.

  Could not find a package configuration file provided by "Torch" with any of
  the following names:

    TorchConfig.cmake
    torch-config.cmake

  Add the installation prefix of "Torch" to CMAKE_PREFIX_PATH or set
  "Torch_DIR" to a directory containing one of the above files.  If "Torch"
  provides a separate development package or SDK, be sure it has been
  installed.
-- Configuring incomplete, errors occurred!
See also "/root/libtorch/mnist/build/CMakeFiles/CMakeOutput.log".

(8)I researched about this error and I modified the CMakeLists.txt as below.   
　I added set(CMAKE_PREFIX_PATH "/libtorch/share/cmake/Torch")　and
   I delted find_package(Torch REQUIRED)

cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
project(mnist)

※set(CMAKE_PREFIX_PATH "/libtorch/share/cmake/Torch")
※ Delete  find_package(Torch REQUIRED)

option(DOWNLOAD_MNIST "Download the MNIST dataset from the internet" ON)
if (DOWNLOAD_MNIST)
  message(STATUS "Downloading MNIST dataset")
  execute_process(
    COMMAND python ${CMAKE_CURRENT_LIST_DIR}/../tools/download_mnist.py
      -d ${CMAKE_BINARY_DIR}/data
    ERROR_VARIABLE DOWNLOAD_ERROR)
  if (DOWNLOAD_ERROR)
    message(FATAL_ERROR "Error downloading MNIST dataset: ${DOWNLOAD_ERROR}")
  endif()
endif()

add_executable(mnist mnist.cpp)
target_compile_features(mnist PUBLIC cxx_range_for)
target_link_libraries(mnist ${TORCH_LIBRARIES})

if (MSVC)
  file(GLOB TORCH_DLLS "${TORCH_INSTALL_PREFIX}/lib/*.dll")
  add_custom_command(TARGET mnist
                     POST_BUILD
                     COMMAND ${CMAKE_COMMAND} -E copy_if_different
                     ${TORCH_DLLS}
                     $<TARGET_FILE_DIR:mnist>)
endif (MSVC)

(8)7)$ cd build and $cmake -DCMAKE_PREFIX_PATH=/libtorch ..  (with New CMakeLists.txt)
 root@460b2c0da7ef : ~ /libtorch/mnist/build# cmake -DCMAKE_PREFIX_PATH=/libtorch ..
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Downloading MNIST dataset
-- Configuring done
-- Generating done
-- Build files have been written to: /root/libtorch/mnist/build
root@460b2c0da7ef: ~ /libtorch/mnist/build# make
Scanning dependencies of target mnist
[ 50%] Building CXX object CMakeFiles/mnist.dir/mnist.cpp.o
/root/libtorch/mnist/mnist.cpp:1:10: fatal error: torch/torch.h: No such file or directory
 #include <torch/torch.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/mnist.dir/build.make:62: recipe for target 'CMakeFiles/mnist.dir/mnist.cpp.o' failed
make[2]: *** [CMakeFiles/mnist.dir/mnist.cpp.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/mnist.dir/all' failed
make[1]: *** [CMakeFiles/mnist.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
root@460b2c0da7ef:~/libtorch/mnist/build# 

(9) I faced 2nd Error  
 No such file or directory  #include <torch/torch.h>

(10) I correct mnist.cpp as blow .
       I changed  C/C++ code from  <torch/toch.h> to  <torch/script.h>

#cahnge 
#include <torch/torch.h> →#include <torch/script.h>

#include <cstddef>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

// Where to find the MNIST dataset.
const char* kDataRoot = "./data";

// The batch size for training.
const int64_t kTrainBatchSize = 64;

// The batch size for testing.
const int64_t kTestBatchSize = 1000;

// The number of epochs to train.
const int64_t kNumberOfEpochs = 10;

// After how many batches to log a new update with the loss value.
const int64_t kLogInterval = 10;

struct Net : torch::nn::Module {
  Net()
      : conv1(torch::nn::Conv2dOptions(1, 10, /*kernel_size=*/5)),
        conv2(torch::nn::Conv2dOptions(10, 20, /*kernel_size=*/5)),
        fc1(320, 50),
        fc2(50, 10) {
    register_module("conv1", conv1);
    register_module("conv2", conv2);
    register_module("conv2_drop", conv2_drop);
    register_module("fc1", fc1);
    register_module("fc2", fc2);
  }

  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));
    x = torch::relu(
        torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
    x = x.view({-1, 320});
    x = torch::relu(fc1->forward(x));
    x = torch::dropout(x, /*p=*/0.5, /*training=*/is_training());
    x = fc2->forward(x);
    return torch::log_softmax(x, /*dim=*/1);
  }

  torch::nn::Conv2d conv1;
  torch::nn::Conv2d conv2;
  torch::nn::FeatureDropout conv2_drop;
  torch::nn::Linear fc1;
  torch::nn::Linear fc2;
};

template <typename DataLoader>
void train(
    size_t epoch,
    Net& model,
    torch::Device device,
    DataLoader& data_loader,
    torch::optim::Optimizer& optimizer,
    size_t dataset_size) {
  model.train();
  size_t batch_idx = 0;
  for (auto& batch : data_loader) {
    auto data = batch.data.to(device), targets = batch.target.to(device);
    optimizer.zero_grad();
    auto output = model.forward(data);
    auto loss = torch::nll_loss(output, targets);
    AT_ASSERT(!std::isnan(loss.template item<float>()));
    loss.backward();
    optimizer.step();

    if (batch_idx++ % kLogInterval == 0) {
      std::printf(
          "\rTrain Epoch: %ld [%5ld/%5ld] Loss: %.4f",
          epoch,
          batch_idx * batch.data.size(0),
          dataset_size,
          loss.template item<float>());
    }
  }
}

template <typename DataLoader>
void test(
    Net& model,
    torch::Device device,
    DataLoader& data_loader,
    size_t dataset_size) {
  torch::NoGradGuard no_grad;
  model.eval();
  double test_loss = 0;
  int32_t correct = 0;
  for (const auto& batch : data_loader) {
    auto data = batch.data.to(device), targets = batch.target.to(device);
    auto output = model.forward(data);
    test_loss += torch::nll_loss(
                     output,
                     targets,
                     /*weight=*/{},
                     at::Reduction::Sum)
                     .template item<float>();
    auto pred = output.argmax(1);
    correct += pred.eq(targets).sum().template item<int64_t>();
  }

  test_loss /= dataset_size;
  std::printf(
      "\nTest set: Average loss: %.4f | Accuracy: %.3f\n",
      test_loss,
      static_cast<double>(correct) / dataset_size);
}

auto main() -> int {
  torch::manual_seed(1);

  torch::DeviceType device_type;
  if (torch::cuda::is_available()) {
    std::cout << "CUDA available! Training on GPU." << std::endl;
    device_type = torch::kCUDA;
  } else {
    std::cout << "Training on CPU." << std::endl;
    device_type = torch::kCPU;
  }
  torch::Device device(device_type);

  Net model;
  model.to(device);

  auto train_dataset = torch::data::datasets::MNIST(kDataRoot)
                           .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
                           .map(torch::data::transforms::Stack<>());
  const size_t train_dataset_size = train_dataset.size().value();
  auto train_loader =
      torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(
          std::move(train_dataset), kTrainBatchSize);

  auto test_dataset = torch::data::datasets::MNIST(
                          kDataRoot, torch::data::datasets::MNIST::Mode::kTest)
                          .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
                          .map(torch::data::transforms::Stack<>());
  const size_t test_dataset_size = test_dataset.size().value();
  auto test_loader =
      torch::data::make_data_loader(std::move(test_dataset), kTestBatchSize);

  torch::optim::SGD optimizer(
      model.parameters(), torch::optim::SGDOptions(0.01).momentum(0.5));

  for (size_t epoch = 1; epoch <= kNumberOfEpochs; ++epoch) {
    train(epoch, model, device, *train_loader, optimizer, train_dataset_size);
    test(model, device, *test_loader, test_dataset_size);
  }
}

(11) I tried to build code.

root@460b2c0da7ef: ~ /libtorch/mnist/build# cmake -DCMAKE_PREFIX_PATH=/libtorch ..
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Downloading MNIST dataset
-- Configuring done
-- Generating done
-- Build files have been written to: /root/libtorch/mnist/build
root@460b2c0da7ef: ~ /libtorch/mnist/build# make
Scanning dependencies of target mnist
[ 50%] Building CXX object CMakeFiles/mnist.dir/mnist.cpp.o
/root/libtorch/mnist/mnist.cpp:1:10: fatal error: torch/script.h: No such file or directory
 #include <torch/script.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/mnist.dir/build.make:62: recipe for target 'CMakeFiles/mnist.dir/mnist.cpp.o' failed
make[2]: *** [CMakeFiles/mnist.dir/mnist.cpp.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/mnist.dir/all' failed
make[1]: *** [CMakeFiles/mnist.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
root@460b2c0da7ef: ~ /libtorch/mnist/build# ^C
root@460b2c0da7ef: ~ /libtorch/mnist/build# 

(12) I could not solve this error .
       Please tell me how do I have to change mnist.cpp and CMakeList.txt.

(13) I cheked Cmake version in Docker container.
     cmake version 3.14.1

(14)My Enviroment
 OS:Ubuntu18.04LTS
 Docker CE
 CUDA10.1
